File: Sources/Sentry/SentrySwizzleWrapper.m
#import "SentrySwizzleWrapper.h"
#import "SentrySwizzle.h"
#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

@implementation SentrySwizzleWrapper

#if SENTRY_HAS_UIKIT
static NSMutableDictionary<NSString *, SentrySwizzleSendActionCallback>
    *sentrySwizzleSendActionCallbacks;
#endif

+ (SentrySwizzleWrapper *)sharedInstance
{
    static SentrySwizzleWrapper *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{ instance = [[self alloc] init]; });
    return instance;
}

+ (void)initialize
{
#if SENTRY_HAS_UIKIT
    if (self == [SentrySwizzleWrapper class]) {
        sentrySwizzleSendActionCallbacks = [NSMutableDictionary new];
    }
#endif
}

#if SENTRY_HAS_UIKIT
- (void)swizzleSendAction:(SentrySwizzleSendActionCallback)callback forKey:(NSString *)key
{
    // We need to make a copy of the block to avoid ARC of autoreleasing it.
    sentrySwizzleSendActionCallbacks[key] = [callback copy];

    if (sentrySwizzleSendActionCallbacks.count != 1) {
        return;
    }

#    pragma clang diagnostic push
#    pragma clang diagnostic ignored "-Wshadow"
    static const void *swizzleSendActionKey = &swizzleSendActionKey;
    SEL selector = NSSelectorFromString(@"sendAction:to:from:forEvent:");
    SentrySwizzleInstanceMethod(UIApplication.class, selector, SentrySWReturnType(BOOL),
        SentrySWArguments(SEL action, id target, id sender, UIEvent * event), SentrySWReplacement({
            [SentrySwizzleWrapper sendActionCalled:action event:event];
            return SentrySWCallOriginal(action, target, sender, event);
        }),
        SentrySwizzleModeOncePerClassAndSuperclasses, swizzleSendActionKey);
#    pragma clang diagnostic pop
}

- (void)removeSwizzleSendActionForKey:(NSString *)key
{
    [sentrySwizzleSendActionCallbacks removeObjectForKey:key];
}

/**
 * For testing. We want the swizzling block above to call a static function to avoid having a block
 * reference to an instance of this class.
 */
+ (void)sendActionCalled:(SEL)action event:(UIEvent *)event
{
    for (SentrySwizzleSendActionCallback callback in sentrySwizzleSendActionCallbacks.allValues) {
        callback([NSString stringWithFormat:@"%s", sel_getName(action)], event);
    }
}

/**
 * For testing.
 */
- (NSDictionary<NSString *, SentrySwizzleSendActionCallback> *)swizzleSendActionCallbacks
{
    return sentrySwizzleSendActionCallbacks;
}
#endif

- (void)removeAllCallbacks
{
#if SENTRY_HAS_UIKIT
    [sentrySwizzleSendActionCallbacks removeAllObjects];
#endif
}
@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Helper/SentrySwizzleWrapper.swift
// swiftlint:disable missing_docs
@_implementationOnly import _SentryPrivate

#if (os(iOS) || os(tvOS) || os(visionOS)) && !SENTRY_NO_UI_FRAMEWORK
import UIKit

@_spi(Private) public typealias SentrySwizzleSendActionCallback = (String, Any?, Any?, UIEvent?) -> Void

@_spi(Private) @objc public class SentrySwizzleWrapper: NSObject {
    
    static var sentrySwizzleSendActionCallbacks = [String: SentrySwizzleSendActionCallback]()
    
    @objc public func swizzleSendAction(_ callback: @escaping SentrySwizzleSendActionCallback, forKey key: String) {
        Self.sentrySwizzleSendActionCallbacks[key] = callback
        SentrySDKLog.debug("Swizzling sendAction for \(key)")

        if Self.sentrySwizzleSendActionCallbacks.count != 1 {
            return
        }

        SentrySwizzleWrapperHelper.swizzle { action, target, sender, event in
            Self.sendActionCalled(action, target: target, sender: sender, event: event)
        }
    }
    
    /// For testing. We want the swizzling block above to call a static function to avoid having a block
    /// reference to an instance of this class.
    static func sendActionCalled(_ action: Selector, target: Any?, sender: Any?, event: UIEvent?) {
        for callback in Self.sentrySwizzleSendActionCallbacks.values {
            callback(String(cString: sel_getName(action)), target, sender, event)
        }
    }

    @objc public func removeSwizzleSendAction(forKey key: String) {
        Self.sentrySwizzleSendActionCallbacks.removeValue(forKey: key)
    }
    
    func removeAllCallbacks() {
        Self.sentrySwizzleSendActionCallbacks.removeAll()
    }
    
    // For test purposes
    static func hasCallbacks() -> Bool {
        return sentrySwizzleSendActionCallbacks.count > 0
    }
}
#endif
// swiftlint:enable missing_docs
